core/src/avm1/value.rs

```diff
@@ -272,14 +272,20 @@ impl<'gc> Value<'gc> {
             (Value::String(a), Value::String(b)) => Ok(*a == *b),
             (Value::Bool(a), Value::Bool(b)) => Ok(a == b),
             (Value::Object(a), Value::Object(b)) => Ok(Object::ptr_eq(*a, *b)),
-            (Value::Object(a), Value::Undefined | Value::Null) => Ok(Object::ptr_eq(
-                *a,
-                activation.context.avm1.global_object_cell(),
-            )),
-            (Value::Undefined | Value::Null, Value::Object(b)) => Ok(Object::ptr_eq(
-                *b,
-                activation.context.avm1.global_object_cell(),
-            )),
+            (Value::Object(a), Value::Undefined | Value::Null)
```

Comment:
After some playing, it looks like the presence of `Object.prototype` is not the real root cause. The `object == number` and `object == string` cases coerce the object into a primitive using `to_primitive_num`, which calls to `valueOf`. The `object == undefined` and `object == null` cases are not different - An object without `Object.prototype` doesn't have `valueOf`, so `to_primitive_num` will return `undefined`.
This also explains why `_global == undefined`, as `_global.valueOf` is not defined.

So, I guess we simply need to merge all cases involving an object. The `object == string` and `object == number` cases are already duplicated, might be worth to merge them as well.

Reply:
Thanks for looking into it! I went ahead and cleaned up this fn and fixed a few other cases:
 * `{valueOf: function() { return NaN; }} == NaN` is true.
 * `Object(1) == Object(1)` is true in SWFv5 but false in SWFv6.
   * SWFv5 always calls `valueOf` at least once on Object params.
   * For Object-Object comparison, it will call `valueOf` once on each, then do a ptr compare.
   * For Object-primitive comparisons, this will call `valueOf` on the object twice.